### PR TITLE
Glsl frontend

### PR DESCRIPTION
**Pull request by JCapucho**
Provide a glsl frontend

features:
- [x] preprocessor
- [X] Global variables
- [ ] Global structs
- [ ] constructors
- [ ] functions (partially implemented)
- [ ] type qualifiers (partially implemented)
-  [X] bindings
- [ ] builtin functions
- [ ] textures and samplers
- [ ] error handling
- [ ] debug information and diagnostics messages
- [ ] `#include` (Not part of the standard but nice to have ???)
- [ ] precision qualifiers
- [ ] Code comments
- [x] CRLF files

### CODE REVIEW

File: src/front/glsl.rs [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+                    crate::TypeInner::Vector { size, kind, width }
+                    | crate::TypeInner::Matrix {
+                        columns: size,
+                        kind,
+                        width,
+                        ..
+                    } => {
+                        const MEMBERS: [char; 4] = ['x', 'y', 'z', 'w'];
+                        if name.len() > 1 {
+                            let mut components = Vec::with_capacity(name.len());
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/front/glsl.rs:621]: I actually copied this from the wgsl implementation but i guess i can add a check for the length of the string and make it a array

[the remark this answers is not in the document]
Comment (kvark) [reply src/front/glsl.rs:621]: oops, nevermind then :) We can do a swipe on the `Vec` use later in the whole codebase.

[the remark this answers is not in the document]
Comment (JCapucho) [reply src/front/glsl.rs:621]: The problem is that expression isn't Default

[the remark this answers is not in the document]
Comment (kvark) [reply src/front/glsl.rs:621]: Yeah. We need to either hook up `ArrayVec`, or just keep a single `Vec` in the parser that we re-use from time to time.

File: src/front/glsl.rs [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::parse_str;
+
+    #[test]
+    fn test_vertex() {
+        let data = r#"#version 450 core
+
+layout(location=0) in vec3 a_position;
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/front/glsl.rs:839]: Funny while moving them to a separate file I noticed one issue CRLF doens't work

File: src/front/wgsl.rs [hunk 3]
```diff
@@ -86,19 +81,19 @@ mod lex {
...
             }
-            '0' ..= '9' => {
-                let (number, rest) = consume_any(input, |c| (c>='0' && c<='9' || c=='.'));
+            '0'..='9' => {
+                let (number, rest) = consume_any(input, |c| (c >= '0' && c <= '9' || c == '.'));
                 (Token::Number(number), rest)
             }
             'a'..='z' | 'A'..='Z' | '_' => {
-                let (word, rest) = consume_any(input, |c| c.is_alphanumeric() || c=='_');
+                let (word, rest) = consume_any(input, |c| c.is_alphanumeric() || c == '_');
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/front/wgsl.rs:89]: Sorry had my editor with format on save I will revert them

File: src/front/glsl.rs [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
+                };
+                let mut value = self.parse_expression(
+                    *value,
+                    expressions,
+                    locals,
+                    locals_map,
+                    parameter_lookup,
+                );
+
+                match op {
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/front/glsl.rs]: Should be fixed

[the remark this answers is not in the document]
Comment (kvark) [reply src/front/glsl.rs]: Github has a "Resolve conversation" button precisely for that

File: Cargo.toml [hunk 5]
```diff
@@ -15,7 +15,10 @@ fxhash = "0.2"
 log = "0.4"
 num-traits = "0.2"
 spirv = { package = "spirv_headers", version = "1" }
-glsl = "4.0.3"
+glsl = { version = "4.0.3", optional = true }
```
Comment (kvark) [inline Cargo.toml:18]: does it have to be "4.0.3" specifically? Unless you require a feature that was introduced in this exact version, let's just put "4" in here

File: Cargo.toml [hunk 6]
```diff
@@ -15,7 +15,10 @@ fxhash = "0.2"
...
+
+[features]
+default = ["glsl"]
```
Comment (kvark) [inline Cargo.toml:21]: let's not have default features

File: test-data/shader.vert [hunk 7]
```diff
@@ -0,0 +1,29 @@
+#version 450 core
```
Comment (kvark) [inline test-data/shader.vert:1]: let's name this shader as something more specific than "shader"

File: src/proc/typifier.rs [hunk 8]
```diff
@@ -67,72 +82,82 @@ impl Typifier {
...
+                        | crate::BinaryOperator::InclusiveOr
+                        | crate::BinaryOperator::ShiftLeftLogical
+                        | crate::BinaryOperator::ShiftRightLogical
+                        | crate::BinaryOperator::ShiftRightArithmetic => self.types[left.index()],
+                    },
                     crate::Expression::Intrinsic { .. } => unimplemented!(),
                     crate::Expression::DotProduct(_, _) => unimplemented!(),
                     crate::Expression::CrossProduct(_, _) => unimplemented!(),
                     crate::Expression::Derivative { .. } => unimplemented!(),
-                    crate::Expression::Call { ref name, ref arguments } => {
```
Comment (kvark) [inline src/proc/typifier.rs:116]: the formatting changes are here still

File: src/front/glsl/helpers.rs [hunk 9]
```diff
@@ -235,6 +236,87 @@ pub fn glsl_to_spirv_type(ty: TypeSpecifierNonArray) -> Option<TypeInner> {
             kind: ScalarKind::Float,
             width: 64,
         },
+        TypeName(ty_name) => {
+            if ty_name.0.contains("texture") {
+                let (base, offset) = match &ty_name.0[..1] {
+                    "t" => {
```
Comment (kvark) [inline src/front/glsl/helpers.rs:242]: nit: looks like you only need an `&str` and `ScalarKind` out of the match and do all the code after it

File: src/front/glsl.rs [hunk 10]
```diff
@@ -0,0 +1,1259 @@
...
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl std::error::Error for Error {}
+
+#[derive(Debug, Copy, Clone)]
+pub enum Global {
```
Comment (kvark) [inline src/front/glsl.rs:147]: does this need to be public?

File: src/front/glsl.rs [hunk 11]
```diff
@@ -0,0 +1,1259 @@
...
+                                    kind: ScalarKind::Sint,
+                                    width: 32,
+                                },
+                            }),
+                        }))
+                    }
+                    "gl_Position" => {
+                        Expression::GlobalVariable(self.globals.fetch_or_append(GlobalVariable {
+                            name: Some(name),
+                            class: StorageClass::Input,
```
Comment (kvark) [inline src/front/glsl.rs:679]: this class should depend on the stage (i.e. it's an output in the VS)

File: src/front/glsl.rs [hunk 12]
```diff
@@ -0,0 +1,1259 @@
...
+                                    kind: ScalarKind::Float,
+                                    width: 32,
+                                },
+                            }),
+                        }))
+                    }
+                    "gl_PointSize" => {
+                        Expression::GlobalVariable(self.globals.fetch_or_append(GlobalVariable {
+                            name: Some(name),
+                            class: StorageClass::Input,
```
Comment (kvark) [inline src/front/glsl.rs:694]: This is `Output`

File: src/front/glsl.rs [hunk 13]
```diff
@@ -0,0 +1,1259 @@
...
+                                    kind: ScalarKind::Float,
+                                    width: 32,
+                                },
+                            }),
+                        }))
+                    }
+                    "gl_ClipDistance" => {
+                        Expression::GlobalVariable(self.globals.fetch_or_append(GlobalVariable {
+                            name: Some(name),
+                            class: StorageClass::Input,
```
Comment (kvark) [inline src/front/glsl.rs:708]: this is `Output`

File: src/front/glsl.rs [hunk 14]
```diff
@@ -0,0 +1,1259 @@
...
+                                    kind: ScalarKind::Float,
+                                    width: 32,
+                                },
+                            }),
+                        }))
+                    }
+                    "gl_CullDistance" => {
+                        Expression::GlobalVariable(self.globals.fetch_or_append(GlobalVariable {
+                            name: Some(name),
+                            class: StorageClass::Input,
```
Comment (kvark) [inline src/front/glsl.rs:722]: this is `Output`

File: src/front/glsl.rs [hunk 15]
```diff
@@ -0,0 +1,1259 @@
...
+                "Unexpected token:\nexpected: {}\ngot: {}",
+                expected
+                    .iter()
+                    .map(|token| {
+                        let mut type_string = token.type_to_string();
+                        type_string.push_str(" |");
+                        type_string
+                    })
+                    .collect::<String>(),
+                got.token.to_string()
```
Comment (kvark) [inline src/front/glsl.rs:77]: aren't tokens implementing `Display` already? Maybe these `to_strings` aren't needed here

File: src/front/glsl/preprocessor.rs [hunk 16]
```diff
@@ -0,0 +1,1730 @@
...
+#[derive(Debug, Copy, Clone)]
+pub enum UnaryOp {
+    Positive,
+    Negative,
+    BitWiseNot,
+    LogicalNot,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum BinaryOp {
```
Comment (kvark) [inline src/front/glsl/preprocessor.rs:69]: do we really need to have a separate enum for this, given that `crate::BinaryOperator` exists?

File: src/front/mod.rs [hunk 17]
```diff
@@ -1,3 +1,5 @@
+#[cfg(feature = "glsl")]
+pub mod glsl;
```
Comment (kvark) [inline src/front/mod.rs:2]: just a thought: if the preprocessor has a lot of complexity and code size, it needs be optional. I'm sure many users will have their own preprocessors anyway (i.e. for `include` handling)

File: src/front/glsl.rs [hunk 18]
```diff
@@ -0,0 +1,1259 @@
...
+    }
+
+    fn parse_global(&mut self, init: InitDeclaratorList) -> Handle<GlobalVariable> {
+        let name = init.head.name.map(|d| d.0);
+        let ty = {
+            let ty = self.parse_type(init.head.ty.ty).unwrap();
+
+            if let Some(specifier) = init.head.array_specifier {
+                match specifier {
+                    ArraySpecifier::Unsized => self.types.fetch_or_append(Type {
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/front/glsl.rs:1086]: This is intended so that things like `var[2][2]` can work

**Pull request merged**